test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/SemaphoreTest.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.testsuite.javalib.util.concurrent
+
+import java.util.concurrent.{Semaphore, TimeUnit}
+import org.junit.Assert._
+import org.junit.Test
+
+class SemaphoreTest {
+
+  @Test def `default constructor should be unfair`(): Unit = {
+    val sem = new Semaphore(1)
+    assertFalse(sem.isFair)
+  }
+
+  @Test def `constructor takes negative values`(): Unit = {
+    val sem = new Semaphore(-1)
+    assertEquals(-1, sem.availablePermits)
+  }
+
+
+  @Test def testDrainNegative(): Unit = {
+    val sem = new Semaphore(-1)
+    sem.drainPermits()
+    assertEquals(0, sem.availablePermits)
+  }
+
+  val acquireMethods: List[Semaphore => Unit] = List(
+    _.drainPermits(),
+    _.tryAcquire(),
+    _.tryAcquire(1),
+    _.tryAcquire(1, 1L, TimeUnit.DAYS),
+    _.tryAcquire(1L, TimeUnit.DAYS),
+  )
+
+  @Test def testAcquires(): Unit = {
+    for (acquire <- acquireMethods) {
+      val sem = new Semaphore(1)
+      acquire(sem)
+      assertEquals(0, sem.availablePermits)
+    }
+  }
+
+  val releaseMethods: List[Semaphore => Unit] = List(
+    _.release(),
+    _.release(1),
+  )
+
+  @Test def testRelease(): Unit = {
+    for (release <- releaseMethods) {
+      val sem = new Semaphore(0)
+      release(sem)
+      assertEquals(1, sem.availablePermits)
+    }
+  }
+
+  @Test def testReleaseThrowsException(): Unit = {
+    val sem = new Semaphore(0)
+    try {
+      sem.release(-1)
+      fail("Exception was not thrown")
+    } catch {
+      case e: Exception => assertTrue(e.isInstanceOf[IllegalArgumentException])
```

Comment:
Match on the exception type directly? (and bubble it up otherwise)?

In general, we have helpers for checking exceptions. You might need to define one in this file, because for the `linker` project we don't have one yet. Look for `assertThrows` in the codebase.